Add ping command for PubSub connections

adrianke opened the pull request:
This allows for periodic checking of the underlying TCP connection.

According to the Redis documentation, PING is one of very few commands that can be issued on a pubsub connection:

"Once the client enters the subscribed state it is not supposed to issue any other commands, except for additional SUBSCRIBE, SSUBSCRIBE, PSUBSCRIBE, UNSUBSCRIBE, SUNSUBSCRIBE, PUNSUBSCRIBE, PING, RESET and QUIT commands"

---

I was not sure if the chosen flow is the best one, ie. sending a serialized PID as part of the payload and using that for sending the reply back to the waiting task.

whatyouhide commented on lib/redix/pubsub/connection.ex:
```diff
@@ -230,6 +230,29 @@ defmodule Redix.PubSub.Connection do
+  def connected({:call, from}, {:ping, _pid}, data) do
+    random_string = :crypto.strong_rand_bytes(12) |> Base.encode64()
+
+    {:ok, task_pid} =
+      Task.start(fn ->
+        res =
+          receive do
+            {:pong, ^random_string} -> :ok
+          after
+            5000 -> :error
+          end
+
+        :ok = :gen_statem.reply(from, res)
+      end)
```
We won't need the task here if we go with a queue of pending clients.

adrianke replied:
Thank you for the feedback.

Would the caller get an `:error` after 5000 ms without the task?

whatyouhide replied:
The flow to implement here is using gen_statem timeouts and processe aliases, see

> Timeout specifies how long to wait for a response. If no response is received within the specified time, this function returns timeout. Assuming that the server executes on a node supporting aliases (introduced in OTP 24) the request will also be abandoned. That is, no response will be received after a time-out. Otherwise, a stray response might be received at a later time.

from the `gen_statem` docs. Basically you don't care about timeouts, you just reply to every pong and BEAM throws away the responses for procs that timed out already

whatyouhide reviewed (commented):
This is gonna be a useful addition. However, few notes:

1. We need to handle `:ping` calls when the gen_statem is in the `disconnected` state.
2. We need tests.
3. Instead of sending a `PING` command with a message, could we just keep a queue of pending `PING` commands, and for every `PONG` that the gen_statem receives we send it to the next caller waiting in the queue? Redis is single-threaded so it will reply to `PING`s in order anyway.

adrianke commented:
I've added some tests, and actually discovered that the reply format is different if you're subscribed to any channels or not.

I've also removed the task, and replaced it with a callback that is sent after 5000 ms.

adrianke commented:
@whatyouhide Just a little reminder about this one. We've run my branch in production at work the last week and haven't encountered any troubles thus far.

whatyouhide commented on lib/redix/pubsub.ex:
```diff
@@ -329,6 +329,17 @@ defmodule Redix.PubSub do
+  @doc """
+  Sends a ping and waits for a reply from the server.
+
+  Upon a successful reply from the server, :ok will be returned. If no reply is received within 5 seconds, :error will be returned.
+
```
```suggestion
  Upon a successful reply from the server, returns `:ok`. If no reply is received within the given timeout, returns `:error`.
```

whatyouhide commented on lib/redix/pubsub.ex:
```diff
@@ -329,6 +329,17 @@ defmodule Redix.PubSub do
+
+  """
+  @spec ping(connection()) :: :ok | :error
+  def ping(conn) do
```
We need to make the timeout here configurable, and not hardcoded to 5 seconds. Do this:

```elixir
def ping(conn, timeout \\ 5000) when is_integer(timeout) and timeout >= 0 do
```

and trickle that down to the implementation.

whatyouhide commented on lib/redix/pubsub/connection.ex:
```diff
@@ -194,6 +218,11 @@ defmodule Redix.PubSub.Connection do
   end
 
+  def disconnected({:call, from}, :ping, _data) do
+    reply = :error
+    {:keep_state_and_data, {:reply, from, reply}}
```
```suggestion
    {:keep_state_and_data, {:reply, from, :error}}
```

whatyouhide commented on lib/redix/pubsub/connection.ex:
```diff
@@ -230,6 +259,19 @@ defmodule Redix.PubSub.Connection do
   end
 
+  def connected({:call, from}, :ping, data) do
+    random_string = :crypto.strong_rand_bytes(12) |> Base.encode64()
```
Let's do the same amount of bytes as in UUIDs.

```suggestion
    random_string = Base.encode64(:crypto.strong_rand_bytes(16))
```

whatyouhide commented on lib/redix/pubsub/connection.ex:
```diff
@@ -230,6 +259,19 @@ defmodule Redix.PubSub.Connection do
+  def connected({:call, from}, :ping, data) do
+    random_string = :crypto.strong_rand_bytes(12) |> Base.encode64()
+    pipeline = [["PING", "ping:#{random_string}"]]
+    data.transport.send(data.socket, Enum.map(pipeline, &Protocol.pack/1))
```
Need to handle if this returns something that is not `:ok` here, and disconnect. We do it in other places too.

whatyouhide commented on lib/redix/pubsub/connection.ex:
```diff
@@ -230,6 +259,19 @@ defmodule Redix.PubSub.Connection do
+    pipeline = [["PING", "ping:#{random_string}"]]
+    data.transport.send(data.socket, Enum.map(pipeline, &Protocol.pack/1))
+
+    Process.send_after(self(), {:ping_callback, random_string}, 5000)
```
Don't use `send_after/3` here. Instead, use `:gen_statem` timeout actions. Return a timeout action from the return value of this function and handle the timeout instead of the message later on. Cancel the timeout if you get a ping with `:cancel`. The `:gen_statem` docs are the best resource for this 🙃

whatyouhide commented on lib/redix/pubsub/connection.ex:
```diff
@@ -15,7 +15,8 @@ defmodule Redix.PubSub.Connection do
     subscriptions: %{},
-    monitors: %{}
+    monitors: %{},
+    pings: []
```
I think you want a `:queue` here rather than a list, right?

whatyouhide reviewed (changes requested):
Going in the right direction, fantastic! Left a few more structural comments 🙃

whatyouhide commented:
@adrianke I took this over and backported it in the commit linked above (applying the suggestions I had left here). Thank you! 🫶
